Add IrisModel interface and IrisGPT3_5Service

FreshLlamanade opened the pull request:
Add initial service for interacting with GPT3.5. Something to build off of.

Hialus commented on src/main/java/de/tum/in/www1/artemis/service/iris/IrisGPT3_5Service.java:
```diff
@@ -0,0 +1,202 @@
+import de.tum.in.www1.artemis.domain.iris.IrisSession;
+
+@Service
+public class IrisGPT3_5Service implements IrisModel {
```
I agree with Teamscale here. A JavaDoc comment explaining what this class does would probably make sense.
This would ideally contain some specifics about what sets GPT3.5 apart.

Hialus commented on src/main/java/de/tum/in/www1/artemis/service/iris/IrisGPT3_5Service.java:
```diff
@@ -0,0 +1,202 @@
+     */
+    @Override
+    @Async
+    public CompletableFuture<String> getResponse(IrisSession session) {
```
Just FYI for anyone else:
This method will likely change in the future:
- Return value will be a list of `IrisMessageContent` instead of String since the parsing of this is LLM dependant
- The parameter will likely change to a List of `IrisMessage` for database reasons

These aren't implemented yet, as they both kinda depend on the database stuff, which is a separate PR, so we will do this once we combine multiple PRs together.

Hialus commented on src/main/java/de/tum/in/www1/artemis/service/iris/IrisGPT3_5Service.java:
```diff
@@ -0,0 +1,202 @@
+            String responseBody = restTemplate.postForObject(apiURL.toString(), request, String.class);
+            log.debug("Response body: {}", responseBody);
+
+            Map<?, ?> response = jsonMapper.readValue(responseBody, Map.class);
```
I think you should add DTO types for this. 
This would allow you to automatically convert the JSON response into a Java type, which you can then read normally without having to use `?` as the generic types.
Also, it would be a cleaner solution that allows for easier expansion in the future.

Hialus commented on src/main/java/de/tum/in/www1/artemis/service/iris/IrisGPT3_5Service.java:
```diff
@@ -0,0 +1,202 @@
+     * @return The API request parameters
+     */
+    private Map<String, Object> createRequestParams(List<IrisMessage> conversationHistory) {
+        Map<String, Object> parameters = new HashMap<>();
```
Creating a DTO for this would probably also make sense for readability.
But you could also leave it like this for now.

FreshLlamanade replied:
For now I think it is a small enough set of parameters that readability is not affected.

Hialus commented on src/main/java/de/tum/in/www1/artemis/service/iris/IrisGPT3_5Service.java:
```diff
@@ -0,0 +1,202 @@
+     */
+    @Value("${artemis.iris.models.gpt3_5.url}")
+    private URL apiURL;
+    // https://ase-eu01.openai.azure.com/openai/deployments/gpt-35/chat/completions?api-version=2023-03-15-preview
```
This should probably not be here.

Hialus reviewed (changes requested):
Thanks for implementing this!
I went through your code again and left some comments for improvements.

FreshLlamanade commented:
@Hialus Thanks for your feedback. I've pushed some changes.

SolizerCodes commented:
I looked over the code, I can't tell if the Spring standards are followed since I am relatively new to Spring, but the rest looks good.
I tested the code and I couldn't get GTP-3.5 to return a message when I changed the "Hello, are you there?". Is is related to the implementation or did I do something wrong?

SolizerCodes reviewed (commented):
I looked over the code, I can't tell if the Spring standards are followed since I am relatively new to Spring, but the rest looks good.
I tested the code and I couldn't get GTP-3.5 to return a message when I changed the "Hello, are you there?". Is is related to the implementation or did I do something wrong?

bassner reviewed (changes requested):
The E2E tests are currently failing

SabrinaGlatz commented on src/main/java/de/tum/in/www1/artemis/service/iris/IrisGPT3_5Service.java:
```diff
@@ -0,0 +1,200 @@
+     * @param conversationHistory The conversation history to convert.
+     * @return The converted messages.
+     */
+    private static List<Map<String, String>> convertMessages(List<IrisMessage> conversationHistory) {
```
Hey Michael, some method names weren't clear for me when I first read the code. I mean you described everything really good in the comment above it, but when the method is used in other places, it's maybe not obvious what it does. Maybe you could change the method name, so it's clear to what format it converts the messages e.g., convertMessagesToGPT3_5Format or something like that. The method that you call in this function (toRoleMessage) also converts Messages to this GPT3_5 format right? So maybe you could also change that method name, to something like toGPT3_5MessageFormat or something instead of toRoleMessage?

SabrinaGlatz commented on src/main/java/de/tum/in/www1/artemis/service/dto/OpenAIChatResponseDTO.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    public static class Usage {
```
It would be good if you add comments to describe the Usage, Choice, and Message classes.

FreshLlamanade commented:
Thanks for the feedback everyone. I've pushed some changes. The E2E Tests should pass again since I've added a profile to `IrisGPT3_5Service` that must be present for Spring to attempt to create it, so we should not be bothered by missing configuration properties. I also added more documentation and changed some method names to better reflect their purposes.

Hialus reviewed (approved):
Code looks good to me now!

Though your PR title could be a little better. E.g. something like "`Iris`: Add GPT 3.5 service". Would make it more readable. 
But this is ultimately up to you (and Patrick).
